extension/persistence/relational-jdbc/src/main/java/org/apache/polaris/extension/persistence/relational/jdbc/DatasourceOperations.java

```diff
@@ -128,22 +143,21 @@ public <T> void executeSelectOverStream(
       @Nonnull Converter<T> converterInstance,
       @Nonnull Consumer<Stream<T>> consumer)
       throws SQLException {
-    try (Connection connection = borrowConnection();
-        Statement statement = connection.createStatement();
-        ResultSet resultSet = statement.executeQuery(query)) {
-      ResultSetIterator<T> iterator = new ResultSetIterator<>(resultSet, converterInstance);
-      consumer.accept(iterator.toStream());
-    } catch (SQLException e) {
-      throw e;
-    } catch (RuntimeException e) {
-      if (e.getCause() instanceof SQLException) {
-        throw (SQLException) e.getCause();
-      } else {
-        throw e;
-      }
-    } catch (Exception e) {
-      throw new RuntimeException(e);
-    }
+    withRetries(
+        () -> {
+          try (Connection connection = borrowConnection();
+              Statement statement = connection.createStatement();
+              ResultSet resultSet = statement.executeQuery(query)) {
+            ResultSetIterator<T> iterator = new ResultSetIterator<>(resultSet, converterInstance);
+            consumer.accept(iterator.toStream());
+            return null;
+          } catch (RuntimeException e) {
+            if (e.getCause() instanceof SQLException ex) {
+              throw ex;
```

Comment:
Why is it necessary to unwrap `SQLException` here? Who wraps it?

Reply:
consumer can't throw anything except RTE, this comes from the ResultSetIterator which takes a consumer which essentially wraps the SQL Exception to RTE and then here we catch this and unwrap. 

Code pointer : https://github.com/apache/polaris/pull/1487/files#diff-e25c4e9df03172cff89a2b96dc5d166a146a36d17af1376515b0dfb20ceddfd6R51

Reply:
In this case the unwrapping can hardly result in retries (it's a read), but we're losing the original exception information (which is likely useful for trouble-shooting).

So, do we have to unwrap here?

Reply:
reads also does required, I am assuming like the row i we are selecting or is in my sample set does get modified in parallel, nevertheless we also retry on like the weird connection issues like connection lost and reset ? we do need SQL exception. So IMHO we should do unwrap for the retry to work or i can make it generic and handle the if else in the Operation and change signature. 

> we're losing the original exception information

Apologies, didn't fully follow, can you please elaborate, we get SQLException in the iterator wrap that to RTE of SQLException and then extract back the SQLException here, not sure where exactly its lost ?

Reply:
The `RuntimeException` in this case has information about the call chain that lead to the exception. The wrapped `SQLException` has only a sub-set of that information. When trouble-shooting it is critical to have the most complete data about the error.

I'd propose not to unwrap exceptions here, but take root cause exceptions into account in the method that decides whether to retry or not.

Reply:
Agree. This can easily lose important contextual information